### PR TITLE
test(connector-sawtooth-socketio): add functional test, bug fix

### PR DESCRIPTION
**Pull request by outSH**
- Add functional test of all functions from sawtooth-socketio validator.
- Refactor sawtooth validator to allow importing as a module, to simplify the functional test.
- Add stopMonitor to terminate tests easily.
- Allow multiple clients to monitor for blocks.
- Fix parsing of URL from config so that it doesn't depend on trailing slash anymore.

Closes: https://github.com/hyperledger/cactus/issues/2107

Depends on: https://github.com/hyperledger/cactus/pull/2109
Depends on: https://github.com/hyperledger/cactus/pull/2110
Depends on: https://github.com/hyperledger/cactus/pull/2047

Signed-off-by: Michal Bajer <michal.bajer@fujitsu.com>

### CODE REVIEW

**Review by jagpreetsinghsasan — APPROVED** [review]

**Review by izuru0 — APPROVED** [review]
LGTM

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:5]: @outSH is `index.js` a typo?

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json
Comment (outSH) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json]: @petermetz Yes it was - fixed

**Comment by outSH** [comment]
@petermetz Rebased and tested with current main - please review

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:5]: @outSH I recommend keeping the conventions of the other packages when it comes to declaring the paths & configuring the compiler (tsconfig.json file). That will avoid a big chunk of the issues like when some packages do not get published due to strange bugs: https://github.com/hyperledger/cactus/issues/2069 or other kinds of automation scripts failing if they make assumptions that some of these paths are uniform everywhere. If there is a specific reason for not going uniform with the package configuration properties then that's a different question of course but if it's all the same to you then I'd highly recommend changing it to match the rest of the packages.

The same goes for the tsconfig.json file, the include paths there ideally would match the rest of the tsconfig.json files which would lead to the compiler output being structured the same way as well.

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json
Comment (outSH) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:5]: BTW the same applies to all `*-socketio-*` connectors, so maybe it is better to handle in separate PR

Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-sawtooth-socketio/package.json:5]: > @petermetz Sorry, I don't understand, can you clarify the conventions you want to keep (would be even nice to have them documented somewhere)? Are the paths supposed to be like `dist/lib/main/typescript/index.js`? If so, it doesn't seem much work and I could do it in this PR I suppose.

@outSH Yeah, that's one of them, but in general every single property in the package.json file that does not have a specific reason (like contributor list or package name/description) to be different should be the same, uniform configuration (same for tsconfig.json files).

Agreed 100% that this should be documented. Best I could do earlier is to make the changes myself last year but then all of that got deleted by others from my own PR so in the end we got what we got.
I'll create a task in the tracker to have a custom check that enforces a set of rules for these files. That way it's not just a markdown file that will almost nobody ever read but instead people will get a failing CI check if anything is wrong with their packages. This will also make our review process more scalable as we continue to add more packages in the future.

---

> BTW the same applies to all *-socketio-* connectors, so maybe it is better to handle in separate PR

 Yeah I'm slowly but surely trying to get us to a state where all packages are as uniform as possible in terms of build configuration to make everybody's lives easier and this would be a part of that so thank you very much for offering. I'm totally fine with doing a separate and bigger PR with all of them, so I'll approve this one as-is and leave it up to you how you want to do it for this specific package. Thank you again!

**Comment by outSH** [comment]
> Yeah I'm slowly but surely trying to get us to a state where all packages are as uniform as possible in terms of build configuration to make everybody's lives easier and this would be a part of that so thank you very much for offering. I'm totally fine with doing a separate and bigger PR with all of them, so I'll approve this one as-is and leave it up to you how you want to do it for this specific package. Thank you again!

@petermetz I've created a task for it here - https://github.com/hyperledger/cactus/issues/2216. I think I'll take care of it after following PR are merged (to limit the merging headaches):
- https://github.com/hyperledger/cactus/pull/2071
- https://github.com/hyperledger/cactus/pull/2089
- https://github.com/hyperledger/cactus/pull/2111

So please merge this when possible, thanks :)

**Comment by petermetz** [comment]
> > Yeah I'm slowly but surely trying to get us to a state where all packages are as uniform as possible in terms of build configuration to make everybody's lives easier and this would be a part of that so thank you very much for offering. I'm totally fine with doing a separate and bigger PR with all of them, so I'll approve this one as-is and leave it up to you how you want to do it for this specific package. Thank you again!
> 
> @petermetz I've created a task for it here - #2216. I think I'll take care of it after following PR are merged (to limit the merging headaches):
> 
> * [feat(fabric-socketio-connector): sending transactions signed on the client-side #2071](https://github.com/hyperledger/cactus/pull/2071)
> * [chore(connector-go-eth): upgrade web3.js to recent version #2089](https://github.com/hyperledger/cactus/pull/2089)
> * [test(connector-sawtooth-socketio): add functional test, bug fix #2111](https://github.com/hyperledger/cactus/pull/2111)
> 
> So please merge this when possible, thanks :)

@outSH Got it, I'll merge this today then. Cheers!

**Pull request merged**
